plan/plan_loader_test.go

```diff
@@ -1,16 +1,17 @@
 // Copyright 2022 Block, Inc.
 
-package plan_test
+package plan
```

Comment:
It's best to avoid this: `pkg_test` helps enforce clean design, clean interfaces/APIs.

What we really want to test is the actual code calls that indirectly cause this bug, so `ChangePan` iirc? Or maybe just something like `got := Sort(Sort(plan))`.

As this test stands now, it's really just testing `deepcopyPlan`, which is fine as an isolated unit test. But what we really want to know is: can we fetch a plan, sort it, then fetch same plan again, sort it again, and get a good 2nd plan?

Reply:
I definitely see the value in enforcing tests that are not dependent on understanding internal implementations of the objects at test. However, and for this specific package, it can be difficult to setup test conditions to properly test, and certainly difficult to keep the test code as concise. I looked at the code, and to test the `Plan` method, we have to load the plans via invoking `LoadShared` which also requires us to have knowledge of the internal logic of the methods and structures at play in order to setup the conditions to make `Plan` testable, not to mention quite a bit more boiler plate testing code (more things that we have to maintain in the future). Thoughts?

In regards to testing the function using `Sort` or `ChangePlan`. I can see value in those packages. But it certainly does not belong here as it creates unnecessary coupling across package boundaries.

IMO, unit test within specific packages need to be concise with that they are testing for and should not have knowledge on the specifics of what callers from the outside may use it for. The end behavior we want is the same, we want to return a deep copy of `Plan` because we don't want a caller to unintentionally modify the data `PlanLoader` is managing. Essentially the gist is the same as the above in that minimizing coupling across package boundaries will improve the maintainability of the code base.